File: frontend/src/lib/reconciliation.ts
/**
 * Tipos + queryKey da Conferência por decisões (Frente G, ADR-067).
 *
 * Espelham `app/schemas/reconciliation.py` — endpoints `GET/POST
 * /processes/{id}/staging-decisions`. Em arquivo próprio (não em
 * `DecisoesPanel.tsx`) porque um arquivo de componente só pode exportar
 * componentes (react-refresh/only-export-components) — `ConsolidacaoPanel`
 * também precisa da mesma queryKey para saber quais staging_ids já viraram
 * decisão.
 */

export interface Evidencia {
  staging_id: number | null;
  documento_id: number | null;
  documento_tipo: string | null;
  campo: string | null;
  valor_bruto: unknown;
  valor_normalizado: unknown;
  unidade: string | null;
  vigencia: string | null;
  status: string;
  fonte_autoritativa: boolean;
}

export interface Decisao {
  chave: { entidade: string; identificador: string; aspecto: string };
  chave_str: string;
  label: string;
  evidencias: Evidencia[];
  concordancia: 'concordam' | 'divergem' | 'fonte_unica';
  nivel_divergencia: 'informativo' | 'atencao' | 'alto' | 'critico' | null;
  delta: number | null;
  percentual: number | null;
  valor_proposto: unknown;
  fonte_autoritativa_doc: string | null;
  estado: 'pendente' | 'decidida' | 'gravada';
  staging_ids: number[];
}

export interface ReconciliationData {
  decisoes: Decisao[];
  sem_agrupamento: Array<{ staging_id: number; motivo: string | null }>;
  total_staging: number;
}

export function decisoesQueryKey(processId: number) {
  return ['staging-decisions', processId];
}

/**
 * STATE-001/ADR-068 (Frente H) — a MESMA chave que `ConsolidacaoPanel.tsx`
 * usa para o banner canônico de progresso da Conferência. Compartilhada aqui
 * para que `DecisoesPanel.tsx` (onde o consultor decide de fato) também
 * invalide o número — sem isto, decidir por `DecisoesPanel` refresca a lista
 * mas deixa o banner acima dela com a contagem velha: dois números sobre o
 * mesmo fato discordando na MESMA tela, o próprio sintoma que STATE-001
 * existe para matar (achado do code review desta frente).
 */
export function progressoConferenciaKey(processId: number) {
  return ['progresso-conferencia', processId];
}

File: app/schemas/reconciliation.py
"""Schemas Pydantic v2 da Conferência por decisões (Frente G, ADR-067).

Espelham `app.services.reconciliation_decisions` (dataclasses puras) — só
serialização, nenhuma regra aqui.
"""

from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict


class ChaveNaturalOut(BaseModel):
    entidade: str
    identificador: str
    aspecto: str

    model_config = ConfigDict(from_attributes=True)


class EvidenciaOut(BaseModel):
    staging_id: Optional[int] = None
    documento_id: Optional[int] = None
    documento_tipo: Optional[str] = None
    campo: Optional[str] = None
    valor_bruto: Optional[Any] = None
    valor_normalizado: Optional[Any] = None
    unidade: Optional[str] = None
    vigencia: Optional[str] = None
    tipo_observacao: Optional[str] = None
    status: str
    fonte_autoritativa: bool = False


class DecisaoOut(BaseModel):
    chave: ChaveNaturalOut
    chave_str: str
    label: str
    evidencias: list[EvidenciaOut] = []
    concordancia: str  # concordam | divergem | fonte_unica
    nivel_divergencia: Optional[str] = None  # informativo|atencao|alto|critico
    delta: Optional[float] = None
    percentual: Optional[float] = None
    valor_proposto: Optional[Any] = None
    fonte_autoritativa_doc: Optional[str] = None
    estado: str  # pendente | decidida | parcialmente_gravada | gravada
    staging_ids: list[int] = []


class SemAgrupamentoOut(BaseModel):
    staging_id: int
    motivo: Optional[str] = None
    target_entity: Optional[str] = None
    target_field: Optional[str] = None
    field_name: Optional[str] = None


class ReconciliationOut(BaseModel):
    decisoes: list[DecisaoOut] = []
    sem_agrupamento: list[SemAgrupamentoOut] = []
    total_staging: int = 0


class DecisaoRequest(BaseModel):
    """Corpo do POST que decide uma decisão agrupada (Frente G, §11: aceite em
    bloco não implementado — só uma decisão por chamada)."""

    entidade: str
    identificador: str
    aspecto: str
    acao: Literal["aceitar", "reabrir", "escolher_fonte", "editar", "reclassificar"]
    staging_id: Optional[int] = None
    valor: Optional[Any] = None
    tipo_observacao: Optional[str] = None
